### DISCUSSION TITLE
AddressEmptyCode - MerkleTree

### DISCUSSION THREAD

**Discussion by ABFX15** [Cyfrin/foundry-full-course-cu#2610] [General]
Hi,

I am struggling to find a resolution for this error when trying to test the MerkleTree airdrop

I'll paste in my merkleairdrop.sol and merkleairdroptest.t.sol and the error 

```
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.24;

import { MerkleProof } from "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
import { IERC20, SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
import { EIP712 } from "@openzeppelin/contracts/utils/cryptography/EIP712.sol";
import { SignatureChecker } from "@openzeppelin/contracts/utils/cryptography/SignatureChecker.sol";
import { ECDSA } from "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
import { MessageHashUtils } from "@openzeppelin/contracts/utils/cryptography/MessageHashUtils.sol";

/**
 * @title Merkle Airdrop - Airdrop tokens to users who can prove they are in a merkle tree
 * @author Cryptab
 */
contract MerkleAirdrop is EIP712 {
    using ECDSA for bytes32;
    using SafeERC20 for IERC20; // Prevent sending tokens to recipients who can’t receive

    error MerkleAirdrop__InvalidProof();
    error MerkleAirdrop__AlreadyClaimed();
    error MerkleAirdrop__InvalidSignature();

    IERC20 private immutable i_airdropToken;
    bytes32 private immutable i_merkleRoot;
    mapping(address => bool) private s_hasClaimed;

    bytes32 private constant MESSAGE_TYPEHASH = keccak256("AirdropClaim(address account,uint256 amount)");

    // define the message hash struct
    struct AirdropClaim {
        address account;
        uint256 amount;
    }

    event Claimed(address account, uint256 amount);
    event MerkleRootUpdated(bytes32 newMerkleRoot);

    /*//////////////////////////////////////////////////////////////
                               FUNCTIONS
    //////////////////////////////////////////////////////////////*/
    constructor(bytes32 merkleRoot, IERC20 airdropToken) EIP712("Merkle Airdrop", "1.0.0") {
        i_merkleRoot = merkleRoot;
        i_airdropToken = airdropToken;
    }

    // claim the airdrop using a signature from the account owner
    function claim(
        address account,
        uint256 amount,
        bytes32[] calldata merkleProof,
        uint8 v,
        bytes32 r,
        bytes32 s
    )
        external
    {
        if (s_hasClaimed[account]) {
            revert MerkleAirdrop__AlreadyClaimed();
        }

        // Verify the signature
        if (!_isValidSignature(account, getMessageHash(account, amount), v, r, s)) {
            revert MerkleAirdrop__InvalidSignature();
        }

        // Verify the merkle proof
        // calculate the leaf node hash
        bytes32 leaf = keccak256(bytes.concat(keccak256(abi.encode(account, amount))));
        // verify the merkle proof (TODO: understand verify)
        if (!MerkleProof.verify(merkleProof, i_merkleRoot, leaf)) {
            revert MerkleAirdrop__InvalidProof();
        }

        s_hasClaimed[account] = true; // prevent users claiming more than once and draining the contract
        emit Claimed(account, amount);
        // transfer the tokens
        i_airdropToken.safeTransfer(account, amount);
    }

    // message we expect to have been signed
    function getMessageHash(address account, uint256 amount) public view returns (bytes32) {
        return _hashTypedDataV4(
            keccak256(abi.encode(MESSAGE_TYPEHASH, AirdropClaim({ account: account, amount: amount })))
        );
    }

    /*//////////////////////////////////////////////////////////////
                             VIEW AND PURE
    //////////////////////////////////////////////////////////////*/
    function getMerkleRoot() external view returns (bytes32) {
        return i_merkleRoot;
    }

    function getAirdropToken() external view returns (IERC20) {
        return i_airdropToken;
    }

    /*//////////////////////////////////////////////////////////////
                             INTERNAL
    //////////////////////////////////////////////////////////////*/

    // verify whether the recovered signer is the expected signer/the account to airdrop tokens for
    function _isValidSignature(
        address signer,
        bytes32 digest,
        uint8 _v,
        bytes32 _r,
        bytes32 _s
    )
        internal
        pure
        returns (bool)
    {
        // could also use SignatureChecker.isValidSignatureNow(signer, digest, signature)
        (
            address actualSigner,
            /*ECDSA.RecoverError recoverError*/
            ,
            /*bytes32 signatureLength*/
        ) = ECDSA.tryRecover(digest, _v, _r, _s);
        return (actualSigner == signer);
    }
    ```
    
    Test Script
    ```
    // SPDX-License-Identifier: MIT
pragma solidity ^0.8.24;

import {MerkleAirdrop} from "../../src/MerkleAirdrop.sol";
import {PenguinToken} from "../../src/PenguinToken.sol";
import {Test} from "forge-std/Test.sol";
import {console} from "forge-std/console.sol";
import {DeployMerkleAirdrop} from "../../script/DeployMerkleAirdrop.s.sol";
import {ZkSyncChainChecker} from "foundry-devops/src/ZkSyncChainChecker.sol";

contract MerkleAirdropTest is ZkSyncChainChecker, Test {
    MerkleAirdrop airdrop;
    PenguinToken token;
    address gasPayer;
    address user;
    uint256 userPrivKey;

    bytes32 merkleRoot = 0xaa5d581231e596618465a56aa0f5870ba6e20785fe436d5bfb82b08662ccc7c4;
    uint256 amountToCollect = (25 * 1e18); // 25.000000
    uint256 amountToSend = amountToCollect * 4;

    bytes32 proofOne = 0x0fd7c981d39bece61f7499702bf59b3114a90e66b51ba2c53abdf7b62986c00a;
    bytes32 proofTwo = 0xe5ebd1e1b5a5478a944ecab36a9a954ac3b6b8216875f6524caa7a1d87096576;
    bytes32[] proof = [proofOne, proofTwo];

    function setUp() public {
        if (!isZkSyncChain()) {
            DeployMerkleAirdrop deployer = new DeployMerkleAirdrop();
            (airdrop, token) = deployer.deployMerkleAirdrop();
        } else {
            token = new PenguinToken();
            airdrop = new MerkleAirdrop(merkleRoot, token);
            token.mint(token.owner(), amountToSend);
            token.transfer(address(airdrop), amountToSend);
        }
        gasPayer = makeAddr("gasPayer");
        (user, userPrivKey) = makeAddrAndKey("user");
    }

    function signMessage(uint256 privKey, address account) public view returns (uint8 v, bytes32 r, bytes32 s) {
        bytes32 hashedMessage = airdrop.getMessageHash(account, amountToCollect);
        (v, r, s) = vm.sign(privKey, hashedMessage);
    }

    function testUsersCanClaim() public {
        uint256 startingBalance = token.balanceOf(user);

        // get the signature
        vm.startPrank(user);
        (uint8 v, bytes32 r, bytes32 s) = signMessage(userPrivKey, user);
        vm.stopPrank();

        // gasPayer claims the airdrop for the user
        vm.prank(gasPayer);
        airdrop.claim(user, amountToCollect, proof, v, r, s);
        uint256 endingBalance = token.balanceOf(user);
        console.log("Ending balance: %d", endingBalance);
        assertEq(endingBalance - startingBalance, amountToCollect);
    }
}
```

Error:
forge script script/Interactions.s.sol:ClaimAirdrop --rpc-url http://localhost:8545 --private-key 0x59c6995e998f97a5a0044966f0945389dc9e86dae88c7a8412f4603b6b78690d --broadcast
[⠢] Compiling...
No files changed, compilation skipped
Traces:
  [1315224] → new ClaimAirdrop@0x5b73C5498c1E3b4dbA84de0F1833c4a029d90519
    └─ ← [Return] 5349 bytes of code

  [1429760] ClaimAirdrop::run()
    ├─ [0] VM::readDir("./broadcast", 3) [staticcall]
    │   └─ ← [Return] [("", "/Users/adambryant/merkle-airdrop/broadcast/DeployMerkleAirdrop.s.sol", 1, true, false), ("", "/Users/adambryant/merkle-airdrop/broadcast/DeployMerkleAirdrop.s.sol/31337", 2, true, false), ("", "/Users/adambryant/merkle-airdrop/broadcast/DeployMerkleAirdrop.s.sol/31337/run-1727979780.json", 3, false, false), ("", "/Users/adambryant/merkle-airdrop/broadcast/DeployMerkleAirdrop.s.sol/31337/run-latest.json", 3, false, false)]
    ├─ [0] VM::toString(31337 [3.133e4]) [staticcall]
    │   └─ ← [Return] "31337"
    ├─ [0] VM::toString(31337 [3.133e4]) [staticcall]
    │   └─ ← [Return] "31337"
    ├─ [0] VM::toString(31337 [3.133e4]) [staticcall]
    │   └─ ← [Return] "31337"
    ├─ [0] VM::readFile("/Users/adambryant/merkle-airdrop/broadcast/DeployMerkleAirdrop.s.sol/31337/run-1727979780.json") [staticcall]
    │   └─ ← [Return] <file>
    ├─ [0] VM::parseJsonUint("<stringified JSON>", ".timestamp") [staticcall]
    │   └─ ← [Return] 1727979780 [1.727e9]
    ├─ [0] VM::toString(0) [staticcall]
    │   └─ ← [Return] "0"
    ├─ [0] VM::keyExistsJson("<JSON file>", "$.transactions[0]") [staticcall]
    │   └─ ← [Return] true
    ├─ [0] VM::toString(0) [staticcall]
    │   └─ ← [Return] "0"
    ├─ [0] VM::keyExistsJson("<JSON file>", "$.transactions[0].contractName") [staticcall]
    │   └─ ← [Return] true
    ├─ [0] VM::parseJsonString("<stringified JSON>", "$.transactions[0].contractName") [staticcall]
    │   └─ ← [Return] "PenguinToken"
    ├─ [0] VM::toString(1) [staticcall]
    │   └─ ← [Return] "1"
    ├─ [0] VM::keyExistsJson("<JSON file>", "$.transactions[1]") [staticcall]
    │   └─ ← [Return] true
    ├─ [0] VM::toString(1) [staticcall]
    │   └─ ← [Return] "1"
    ├─ [0] VM::keyExistsJson("<JSON file>", "$.transactions[1].contractName") [staticcall]
    │   └─ ← [Return] true
    ├─ [0] VM::parseJsonString("<stringified JSON>", "$.transactions[1].contractName") [staticcall]
    │   └─ ← [Return] "MerkleAirdrop"
    ├─ [0] VM::toString(1) [staticcall]
    │   └─ ← [Return] "1"
    ├─ [0] VM::parseJsonAddress("<stringified JSON>", "$.transactions[1].contractAddress") [staticcall]
    │   └─ ← [Return] 0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512
    ├─ [0] VM::toString(2) [staticcall]
    │   └─ ← [Return] "2"
    ├─ [0] VM::keyExistsJson("<JSON file>", "$.transactions[2]") [staticcall]
    │   └─ ← [Return] true
    ├─ [0] VM::toString(2) [staticcall]
    │   └─ ← [Return] "2"
    ├─ [0] VM::keyExistsJson("<JSON file>", "$.transactions[2].contractName") [staticcall]
    │   └─ ← [Return] true
    ├─ [0] VM::parseJsonString("<stringified JSON>", "$.transactions[2].contractName") [staticcall]
    │   └─ ← [Return] "PenguinToken"
    ├─ [0] VM::toString(3) [staticcall]
    │   └─ ← [Return] "3"
    ├─ [0] VM::keyExistsJson("<JSON file>", "$.transactions[3]") [staticcall]
    │   └─ ← [Return] true
    ├─ [0] VM::toString(3) [staticcall]
    │   └─ ← [Return] "3"
    ├─ [0] VM::keyExistsJson("<JSON file>", "$.transactions[3].contractName") [staticcall]
    │   └─ ← [Return] true
    ├─ [0] VM::parseJsonString("<stringified JSON>", "$.transactions[3].contractName") [staticcall]
    │   └─ ← [Return] "PenguinToken"
    ├─ [0] VM::toString(4) [staticcall]
    │   └─ ← [Return] "4"
    ├─ [0] VM::keyExistsJson("<JSON file>", "$.transactions[4]") [staticcall]
    │   └─ ← [Return] false
    ├─ [0] VM::toString(31337 [3.133e4]) [staticcall]
    │   └─ ← [Return] "31337"
    ├─ [0] VM::readFile("/Users/adambryant/merkle-airdrop/broadcast/DeployMerkleAirdrop.s.sol/31337/run-latest.json") [staticcall]
    │   └─ ← [Return] <file>
    ├─ [0] VM::parseJsonUint("<stringified JSON>", ".timestamp") [staticcall]
    │   └─ ← [Return] 1727979780 [1.727e9]
    ├─ [0] VM::startBroadcast()
    │   └─ ← [Return] 
    ├─ [33437] 0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512::claim(0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266, 25000000000000000000 [2.5e19], [0xd1445c931158119b00449ffcac3c947d028c0c359c34a6646d95962b3b55c6ad, 0xe5ebd1e1b5a5478a944ecab36a9a954ac3b6b8216875f6524caa7a1d87096576], 28, 0xfbd2270e6f23fb5fe9248480c0f4be8a4e9bd77c3ad0b1333cc60b5debc51160, 0x2a2a06c24085d8d7c038bad84edc53664c8ce0346caeaa3570afec0e61144dc1)
    │   ├─ [3000] PRECOMPILES::ecrecover(0x184e30c4b19f5e304a89352421dc50346dad61c461e79155b910e73fd856dc72, 28, 113901832435254568846590346974893520100064630318508576867858738096657341419872, 19071393864720662770355241039489023494212240407929980452046490912886082522561) [staticcall]
    │   │   └─ ← [Return] 0x000000000000000000000000f39fd6e51aad88f6f4ce6ab8827279cfffb92266
    │   ├─ emit Claimed(account: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266, amount: 25000000000000000000 [2.5e19])
    │   ├─ [0] 0x0000000000000000000000000000000000000000::transfer(0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266, 25000000000000000000 [2.5e19])
    │   │   └─ ← [Stop] 
    │   └─ ← [Revert] AddressEmptyCode(0x0000000000000000000000000000000000000000)
    └─ ← [Revert] AddressEmptyCode(0x0000000000000000000000000000000000000000)

Error: 
script failed: AddressEmptyCode(0x0000000000000000000000000000000000000000)

**Comment by EngrPips** [comment]
Hello @ABFX15, Please show your `ClaimAirdrop` script

**Reply by ABFX15** [reply]
All good, got this working!

**Reply by EngrPips** [reply]
Awesome, Keep crushing it.

**Reply by vegazeronos** [reply]
hi @ABFX15 , can you share your solution? i got the same problem tho

here is my DeployMerkleAirdrop.s.sol script
```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.24;

import {Script} from "forge-std/Script.sol";
import {ShibToken} from "src/ShibToken.sol";
import {MerkleAirdrop} from "src/MerkleAirdrop.sol";
import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";

contract DeployMerkleAirdrop is Script {
    bytes32 private s_merkleRoot = 0xaa5d581231e596618465a56aa0f5870ba6e20785fe436d5bfb82b08662ccc7c4;
    uint256 private constant AMOUNT_TO_MINT = 4 * 25 * 1e18;
    uint256 private constant AMOUNT_TO_TRANSFER = 25 * 1e18;

    function _deployMerkleAirdropContracts() public returns (MerkleAirdrop, ShibToken) {
        vm.startBroadcast();
        ShibToken token = new ShibToken();
        MerkleAirdrop airdrop = new MerkleAirdrop(s_merkleRoot, IERC20(token));
        token.mint(token.owner(), AMOUNT_TO_MINT);
        IERC20(token).transfer(address(airdrop), AMOUNT_TO_TRANSFER);
        vm.stopBroadcast();
        return (airdrop, token);
    }

    function run() external returns (MerkleAirdrop, ShibToken) {
        _deployMerkleAirdropContracts();
    }
}
```

**Reply by EngrPips** [reply]
Hello @vegazeronos, Please show your script process stack trace.

**Reply by vegazeronos** [reply]
sorry @EngrPips , its my dumbdumb brain, i dont return the "_deployMeklreAirdropContracts()" and now its already deploy well... however, i've got new proble, i think i will post it on a new discussion tho

**Reply by EngrPips** [reply]
Yeah sure, post new problem in new discussion. Thank you.
